docs/proposals/2022-07-13-appset-progressive-rollout-strategy.md

```diff
@@ -0,0 +1,169 @@
+---
+title: ApplicationSet Progressive Rollout Strategy
+authors:
+  - "@wmgroot"
+  - "@cnmcavoy"
+sponsors:
+  - indeed.com
+reviewers:
+  - "@alexmt"
+  - TBD
+approvers:
+  - "@alexmt"
+  - TBD
+
+creation-date: 2022-07-13
+last-updated: 2022-07-13
+---
+
+# ApplicationSet Progressive Rollout Strategy
+
+## Summary
+
+Enhance the ArgoCD ApplicationSet resource to embed a rollout strategy for a progressive application resource update after the ApplicationSet spec or Application templates are modified.
+Further discussion and interest has been communicated here: https://github.com/argoproj/argo-cd/issues/9437
+
+## Motivation
+
+As cluster operators, we would like to make changes to ApplicationSets which may target multiple environments, pre-defined staging areas, or other configurations, and have these changes rolled out in a declarative, defined manner rather than all at once as ApplicationSets currently behave. A progressive ApplicationSet rollout would prevent mistakes in configuration from having a larger blast radius than intended and give cluster operators a chance to verify and have confidence in their changes.
+
+### Goals
+
+Users are able to make a single change to ApplicationSet that is updated across the generated Applications in a controlled manner. When this enhancement is enabled, Applications are updated in a declaractive order, instead of simultaneously.
+
+### Non-Goals
+
+Handling controlled rollouts for changes to a helm chart or raw manifests referenced by the Applications managed by the ApplicationSet. We understand this would be valuable, but we would like to implement the rollout implementation handling only changes to the ApplicationSet initially.
+
+## Proposal
+
+This is where we get down to details of what the proposal is about.
+
+### Use cases
+
+Add a list of detailed use cases this enhancement intends to take care of.
+
+#### Use case 1:
+As a user, I would like to declaratively control the rollout order of ApplicationSet changes to its generated Application resources.
+
+We propose adding a `RollingUpdate` strategy spec (taking inspiration from other controllers).
+
+The rolling update strategy deterministically chooses applications to update one by one, proceeding each step only if the previous application syncs completed successfully, and respecting the maxUpdate value.
```

Comment:
nit: the example covers the behavior of maxUpdate clearly, but I would find it helpful to clarify that `one by one` is an option and not the rule for the proposal.  

So, this could be reworded along the lines of:

The rolling update strategy deterministically chooses applications to update following a maxUpdate value.  If maxUpdate is set to 1, then applications are updated one by one, proceeding each step only if the previous application syncs completed successfully.  If set to more than 1, then applications are updated in parallel up to that number.

Reply:
:+1: